### PR TITLE
Add swift_getTypeByMangledNameInContextInMetadataState such that we can

### PR DESCRIPTION
**Pull request by aschwaighofer**
use getTypeByMangledName when abstract metadata state is requested

This can significantly reduce the code size of apps constructing deeply
nested types with conditional conformances.

Requires a new runtime.

rdar://57157619

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/MetadataRequest.cpp [hunk 1]
```diff
@@ -2198,11 +2211,15 @@ emitMetadataAccessByMangledName(IRGenFunction &IGF, CanType type,
   }
 
   // Get or create a shared helper function to do the instantiation.
+  auto instantiationFnName =
+      request.isStaticallyAbstract()
+          ? "__swift_instantiateConcreteTypeFromMangledNameAbstract"
+          : "__swift_instantiateConcreteTypeFromMangledName";
```
Comment (jckarter) [inline lib/IRGen/MetadataRequest.cpp]: Thanks for implementing this, Arnold! I wonder whether, if we're targeting Swift >= 5.2, we could only emit `__swift_instantiateConcreteTypeFromMangledNameAbstract` and use it for all metadata requests. That would probably mean a couple more instructions at call sites to materialize the metadata request kind argument, though.

File: include/swift/Runtime/RuntimeFunctions.def [hunk 2]
```diff
@@ -1409,6 +1409,20 @@ FUNCTION(GetTypeByMangledNameInContext, swift_getTypeByMangledNameInContext,
...
+//                        size_t typeNameLength,
+//                        const TargetContextDescriptor<InProcess> *context,
+//                        const void * const *genericArgs)
+FUNCTION(GetTypeByMangledNameInContextInMetadataState,
+         swift_getTypeByMangledNameInContextInMetadataState, SwiftCC,
+         GetTypesInAbstractMetadataStateAvailability,
+         RETURNS(TypeMetadataPtrTy),
+         ARGS(SizeTy, Int8PtrTy, SizeTy, TypeContextDescriptorPtrTy,
+              Int8PtrPtrTy),
+         ATTRS(NoUnwind, ArgMemOnly))
```
Comment (jckarter) [inline include/swift/Runtime/RuntimeFunctions.def]: It'd also be good to add the `swift_getTypeByMangledNameInEnvironmentInMetadataState` variant, so we have it available for future use to instantiate metadata in generic environments that don't have an associated type context.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test linux

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Pull request merged**